### PR TITLE
CA-249056: Add a unique ID to XenCenter Check For Update

### PR DESCRIPTION
**Pull request by GaborApatiNagy**

### CODE REVIEW

**Comment by stephen-turner** [comment]
What do typical user agents look like in your scheme. Do they correspond to the usual convention from other vendors?

**Comment by GaborApatiNagy** [comment]
@stephen-turner Good point. At the moment it would be: "XenCenter (v.1.2.3)", but I think I should change it to "XenCenter/1.2.3" to make it standard and to comply with the format defined in https://tools.ietf.org/html/rfc7231#section-5.5.3

I'll make this change and update this PR now.

File: XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs [hunk 1]
```diff
@@ -293,6 +294,12 @@ protected virtual XmlDocument FetchCheckForUpdatesXml(string location)
             var xdoc = new XmlDocument();
             var uri = new Uri(location);
             var proxy = XenAdminConfigManager.Provider.GetProxyFromSettings(Connection, false);
+            
+            var uniqueIdHash = GetUniqueIdHash();
+
+            string buildText = Helpers.CommonCriteriaCertificationRelease
+                        ? string.Format("{0}: {1}", Program.Version.Revision, Messages.COMMON_CRITERIA_TEXT)
```
Comment (stephen-turner) [inline XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs]: We don't have any CC versions any more; but even if we did, I don't think we want the COMMON_CRITERIA_TEXT in the user-agent. Let's just lose it.

File: XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs [hunk 2]
```diff
@@ -293,6 +294,12 @@ protected virtual XmlDocument FetchCheckForUpdatesXml(string location)
             var xdoc = new XmlDocument();
             var uri = new Uri(location);
             var proxy = XenAdminConfigManager.Provider.GetProxyFromSettings(Connection, false);
+            
+            var uniqueIdHash = GetUniqueIdHash();
+
+            string buildText = Helpers.CommonCriteriaCertificationRelease
+                        ? string.Format("{0}: {1}", Program.Version.Revision, Messages.COMMON_CRITERIA_TEXT)
+                        : Program.Version.Revision.ToString();
```
Comment (stephen-turner) [inline XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs]: This just takes the build number. I think you want the three-digit version number instead.

File: XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs [hunk 3]
```diff
@@ -313,5 +322,26 @@ protected virtual XmlDocument FetchCheckForUpdatesXml(string location)
 
             return xdoc;
         }
+
+        private string GetUniqueIdHash()
+        {
+            string uniqueIdHash = "nil";
+
+            var managementObj = new System.Management.ManagementObject("Win32_OperatingSystem=@");
```
Comment (stephen-turner) [inline XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs]: I suppose all these system calls are OK, but I can't be sure that none of them can throw exceptions. Maybe we should wrap the whole of lines 330-342 in a try-catch block.

File: XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs [hunk 4]
```diff
@@ -313,5 +322,26 @@ protected virtual XmlDocument FetchCheckForUpdatesXml(string location)
...
+        private string GetUniqueIdHash()
+        {
+            string uniqueIdHash = "nil";
+
+            var managementObj = new System.Management.ManagementObject("Win32_OperatingSystem=@");
+            string serialNumber = (string)managementObj["SerialNumber"];
+
+            if (!string.IsNullOrWhiteSpace(serialNumber))
+            {
+                var serialBytes = Encoding.ASCII.GetBytes("JustSomeSaltForCheckForUpdates" + serialNumber);
```
Comment (stephen-turner) [inline XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs]: I'm not sure what the salt is doing here. I don't see any advantage of a publicly known salt, and it just makes the md5sum longer to compute.

File: XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs [hunk 5]
```diff
@@ -38,6 +38,7 @@
 using XenAdmin.Core;
 using System.Diagnostics;
 using System.Net;
+using System.Text;
```
Comment (stephen-turner) [inline XenAdmin/Actions/GUIActions/DownloadUpdatesXmlAction.cs]: I think we should keep DownloadUpdatesXmlAction in XenModel, and pass in the user agent and user agent id that we want to use, rather than move it all to XenAdmin.

**Comment by stephen-turner** [comment]
Please also step through the code and make sure that everything is behaving as expected.

**Comment by stephen-turner** [comment]
http://stackoverflow.com/questions/3138893/detail-about-win32-operatingsystem-serialnumber and http://stackoverflow.com/questions/14213020/getting-windows-serial-number-was-getting-machineguid-from-registry raise doubts about the uniqueness of SerialNumber. We should consider carefully what level of uniqueness we need, and what call will best provide this.

File: XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs [hunk 6]
```diff
@@ -35,6 +35,7 @@
 using System.Threading;
 using System.IO;
 using XenCenterLib.Archive;
+using System.Management;
```
Comment (stephen-turner) [inline XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs]: Please remove unnecessary change

File: XenModel/Actions/Updates/DownloadUpdatesXmlAction.cs [hunk 7]
```diff
@@ -62,39 +63,44 @@ public List<XenServerVersion> XenServerVersionsForAutoCheck
...
                 }
                 return new List<XenServerVersion>();
             }
         }
 
-        private readonly bool _checkForXenCenter;
-        private readonly bool _checkForServerVersion;
-        private readonly bool _checkForPatches;
-        private readonly string _checkForUpdatesUrl;
+        private readonly bool checkForXenCenter;
```
Comment (stephen-turner) [inline XenModel/Actions/Updates/DownloadUpdatesXmlAction.cs]: Please don't make any changes beyond what is necessary here

File: XenModel/XenModel.csproj [hunk 8]
```diff
@@ -52,6 +52,7 @@
       <RequiredTargetFramework>3.5</RequiredTargetFramework>
     </Reference>
     <Reference Include="System.Drawing" />
+    <Reference Include="System.Management" />
```
Comment (stephen-turner) [inline XenModel/XenModel.csproj]: Please remove unnecessary change

**Comment by stephen-turner** [comment]
Everything looks good on the server side, so merging.

**Comment by GaborApatiNagy** [comment]
Great news, thanks for merging

On 26 Apr 2017 12:46, "Stephen Turner" <notifications@github.com> wrote:

Merged #1537 <https://github.com/xenserver/xenadmin/pull/1537>.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/xenserver/xenadmin/pull/1537#event-1058432643>, or mute
the thread
